### DISCUSSION TITLE
Herokuish vs Dockerfile vs Docker-Image

### DISCUSSION THREAD

**Discussion by guettli** [dokku/dokku#4254] [General]
From the docs:

> Dokku supports deploying applications in a few ways:
> 
> Heroku buildpacks via Herokuish: See the herokuish buildpacks documentation to learn about the different ways to specify a buildpack.
> Dockerfile: See the dockerfile documentation to learn about the different ways to configure Dockerfile-based deploys.
> Docker Image: See the docker image documentation to learn about how to deploy a Docker Image.

I am unsure which one to choose.

I am starting from scratch and never used Heroku before, and I don't plan it at the moment.

What do you think would be the right choice for me?

**Comment by josegonzalez** [comment]
The default method is the first one, and what folks will typically see when they push an app. This does a bunch of fancy autodetection - using more or less the same method as Heroku - to build a docker image.

If you need more control, you can instead create a `Dockerfile` in your repository. Some folks choose this over Heroku Buildpacks because... well they just want to use Dokku as a way to handle the app lifecycle, and not necessary what gets built.

Finally, you can deploy via straight Docker Image. This is nice if you want to separate your build and deploy step - maybe you have some other system elsewhere that must perform scanning before deployments or want to deploy the image you tested in CI - so it's there as an option.

What you choose is kind of up to you, though the auto-detected default (which is shown on the application deployment tutorial) of Heroku Buildpacks works fairly well for most people.

**Reply by schmijos** [reply]
Is it possible to use _dokku_ to build the slug on one server (build server) with _herokuish_, but then move the slug to another server and still run it with _dokku_?

The problem is that building my apps needs a lot of internet access (meaning firewall rules) while running them doesn't.

**Reply by josegonzalez** [reply]
There isn't any "slug" we export, but the new 0.25.x release will have `registry` support, allowing you to push the image somewhere (a docker registry) after a build and then move that to whatever server and deploy it (likely with `git:from-image`). That functionality was originally for alternate schedulers, but can be used for something like this too.

**Comment by guettli** [comment]
Thank you very much. If I were you, I would update the text here and provide above explanation here :https://github.com/dokku/dokku/blob/master/docs/deployment/application-deployment.md

**Comment by cstuder** [comment]
I agree with @guettli, especially deploying Docker images isn't that trivial with the whole tagging stuff.

**Reply by badsyntax** [reply]
Agree that the process for deploying docker images is somewhat awkward, but this is [being worked on](https://github.com/dokku/dokku/issues/4296) and i expect the docs to be updated once that feature is complete.

**Comment by oneCodeScholar** [comment]
I've been working in a docker container and using npm build to create an app. Is it possible to have multiple apps on one Dokku server? I have two dokku apps, each with the same build. I have the ports mapped to 3000 on one and 3001 on the other. When I deploy the app they get built using the 3000 port on both it seems. Running dokku logs for the apps it shows: ready - started server on 0.0.0.0:3000, url: http://localhost:3000    How can I fix this? Or is it best to have only just one app per Dokku server?

**Reply by josegonzalez** [reply]
This seems like a different problem. if you are still having this issue, I would open a new discussion.
